prometheus/src/lib.rs

```diff
@@ -6,40 +6,97 @@ mod snapshot_metrics;
+use identity_info::{map_vote_identity_to_info, IdentityInfoMap};
 use solana_gossip::cluster_info::ClusterInfo;
-use solana_runtime::snapshot_config::SnapshotConfig;
+use solana_runtime::{
+    bank_forks::BankForks, commitment::BlockCommitmentCache, snapshot_config::SnapshotConfig,
+};
 use solana_sdk::pubkey::Pubkey;
-use std::{collections::HashSet, sync::Arc};
+use std::{
+    collections::HashSet,
+    sync::{Arc, RwLock},
+    thread,
+};
 
 #[derive(Clone, Copy)]
 pub struct Lamports(pub u64);
 
-pub fn render_prometheus(
-    banks_with_commitments: BanksWithCommitments,
-    cluster_info: &Arc<ClusterInfo>,
-    vote_accounts: &Arc<HashSet<Pubkey>>,
-    identity_config: &Arc<IdentityInfoMap>,
-    snapshot_config: &Option<SnapshotConfig>,
-) -> Vec<u8> {
-    // There are 3 levels of commitment for a bank:
-    // - finalized: most recent block *confirmed* by supermajority of the
-    // cluster.
-    // - confirmed: most recent block that has been *voted* on by supermajority
-    // of the cluster.
-    // - processed: most recent block.
-    let mut out: Vec<u8> = Vec::new();
-    bank_metrics::write_bank_metrics(&banks_with_commitments, &mut out).expect("IO error");
-    cluster_metrics::write_cluster_metrics(
-        &banks_with_commitments,
-        &cluster_info,
-        vote_accounts,
-        identity_config,
-        &mut out,
-    )
-    .expect("IO error");
-    if let Some(snapshot_config) = snapshot_config {
-        snapshot_metrics::write_snapshot_metrics(snapshot_config, &mut out).expect("IO error");
+pub struct PrometheusMetrics {
+    bank_forks: Arc<RwLock<BankForks>>,
+    block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+    cluster_info: Arc<ClusterInfo>,
+    vote_accounts: Arc<HashSet<Pubkey>>,
+    snapshot_config: Option<SnapshotConfig>,
+    /// Initialized based on vote_accounts. Maps identity
+    /// pubkey associated with the vote account to the validator info.
+    /// Since loading accounts takes a lot of time, we initialize it in a
+    /// separate thread, hence the RwLock - to set the data later from a
+    /// different thread.
+    identity_info_map: RwLock<Option<IdentityInfoMap>>,
+}
+
+impl PrometheusMetrics {
+    pub fn new(
+        bank_forks: Arc<RwLock<BankForks>>,
+        block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+        cluster_info: Arc<ClusterInfo>,
+        vote_accounts: Arc<HashSet<Pubkey>>,
+        snapshot_config: Option<SnapshotConfig>,
+    ) -> Arc<Self> {
+        let prom_metrics = Self {
+            bank_forks: bank_forks.clone(),
+            block_commitment_cache,
+            cluster_info,
+            vote_accounts: vote_accounts.clone(),
+            identity_info_map: RwLock::new(None),
+            snapshot_config,
+        };
+        let prom_metrics = Arc::new(prom_metrics);
+
+        let prom_metrics_clone = prom_metrics.clone();
+        thread::spawn(move || {
+            // TODO: This can panic, we should handle it better
```

Comment:
I meant `expect(..)` in `map_vote_identity_to_info`. The issue is that if it happens it will be hard to catch since the thread will just panic into oblivion.

Reply:
> hard to catch since the thread will just panic into oblivion

Yeah this is a footgun in Rust, I always set `panic = "abort"` in the Cargo config to avoid that, it’s more efficient too, but I’m not sure if we can just enable that for Solana.